Add `auto` option for `config.use_ansi_coloring`

cptpiepmatz opened the pull request:
<!--
if this PR closes one or more issues, you can automatically link the PR with
them by using one of the [*linking keywords*](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), e.g.
- this PR should close #xxxx
- fixes #xxxx

you can also mention related issues, PRs or discussions!
-->

# Description
<!--
Thank you for improving Nushell. Please, check our [contributing guide](../CONTRIBUTING.md) and talk to the core team before making major changes.

Description of your pull request goes here. **Provide examples and/or screenshots** if your changes affect the user experience.
-->

In this PR I continued the idea of #11494, it added an `auto` option to the ansi coloring config option, I did this too but in a more simple approach.

So I added a new enum `UseAnsiColoring` with the three values `True`, `False` and `Auto`. When that value is set to `auto`, the default value, it will use `std::io::stdout().is_terminal()` to decided whether to use ansi coloring. This allows to dynamically decide whether to print ansi color codes or not, [cargo does it the same way](https://github.com/rust-lang/cargo/blob/652623b779c88fe44afede28bf7f1c9c07812511/src/bin/cargo/main.rs#L72). `True` and `False` act as overrides to the `is_terminal` check. So with that PR it is possible to force ansi colors on the `table` command or automatically remove them from the miette errors if no terminal is used.

# User-Facing Changes
<!-- List of all changes that impact the user experience here. This helps us keep track of breaking changes. -->

Terminal users shouldn't be affected by this change as the default value was `true` and `is_terminal` returns for terminals `true` (duh). Non-terminal users, that use `nu` in some embedded way or the engine implemented in some other way (like my jupyter kernel) will now have by default no ansi coloring and need to enable it manually if their environment allows it.

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use toolkit.nu; toolkit test stdlib"` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->

The test for fancy errors expected ansi codes, since tests aren't run "in terminal", the ansi codes got stripped away.
I added a line that forced ansi colors above it. I'm not sure if that should be the case or if we should test against no ansi colors.

- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :green_circle: `toolkit test`
- :green_circle: `toolkit test stdlib`

# After Submitting
<!-- If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date. -->

This should resolve #11464 and partially #11847. This also closes #11494.

fdncred commented:
Seems pretty straight forward. Thanks!

fdncred commented:
please fix the conflicts and we can land this

cptpiepmatz commented:
@fdncred should be fixed

fdncred commented:
@cptpiepmatz Does this support [NO_COLOR](https://no-color.org/). If it does, I missed it in the code. If not, it would be good to add it.

I'm wondering if we should support `FORCE_COLOR` env var as well? I think this crate supports all this. https://docs.rs/supports-color/latest/supports_color/ I'm not saying we should use it, but it may provide inspiration to make this PR more thorough.

cptpiepmatz commented:
@fdncred  I made several changes to support environment variables here:

1. The `get` method now requires an `EngineState` to evaluate environment variables. This ensures that we process these variables for every command, meaning each call to `get` will dynamically evaluate them.

2. With the new config behavior, most users will likely have the value set to `auto` by default. When `auto` is used, it respects the `NO_COLOR`, `FORCE_COLOR`, and `CLICOLOR` environment variables. However, users can still explicitly set the value to `true` or `false` in their config to override these variables. Our internal code does this too, for example, to enforce disabling ANSI coloring.

3. Since `get` became a bit more complex, I reorganized some logic in `table` to reduce unnecessary work.

With this new behavior, we should maybe update the docs and add something in the changelog.

fdncred commented on crates/nu-protocol/src/config/output.rs:
```diff
@@ -24,3 +26,116 @@ impl UpdateFromValue for ErrorStyle {
+}
+
+impl UseAnsiColoring {
+    /// Determines whether ANSI colors should be used.
```
Great documentation here. Thanks for taking the time to spell out exactly what it does.

fdncred commented on crates/nu-protocol/src/value/mod.rs:
```diff
@@ -655,6 +655,35 @@ impl Value {
+            | Value::Int { val: 0, .. }
+            | Value::Float { val: 0.0, .. }
+            | Value::Nothing { .. } => Some(false),
+            Value::String { val, .. } if val.as_str() == "" => Some(false),
```
I'm wondering if we should support a string 0/1 like `"0"` or `"1"`? Or maybe that's covered somewhere/somehow else?

cptpiepmatz replied:
String "0" should definitely also represent a `false` I just missed that one here. Good point.

fdncred reviewed (changes requested):
This is exactly what I was asking for. I have a few more comments. Just let me know what you think.

Also, it would be wonderful to have a couple tests for these env var settings, if you think that's possible? I can't remember if our test harness supports such a thing. I also want to test this myself, which I haven't done yet. 

Thanks so much for working on this. I think it'll be a great addition to nushell.

cptpiepmatz commented:
Since `UseAnsiColoring` needed more lines of code than I planned, I moved it into a nother module. I also added tests for `Value::as_env_bool` and `UseAnsiColoring::get`. I don't think I need to test consumers of the `get` method as it simply returns a boolean.

cptpiepmatz commented:
@fdncred it's green 🎉 

Also imo we should add a label to get this in to the changelog.

fdncred reviewed (commented):
Sorry to keep hitting you with these "one more things" but I've noticed a couple things.

1. I think we need some text in the [doc_config.nu]( https://github.com/nushell/nushell/blob/38694a9850a08f82a9a887c64df55c9f561e4c61/crates/nu-utils/src/default_files/doc_config.nu#L247-L252)
2. In my testing, I've noticed that it mostly works but there are some edge cases. NO_COLOR=1 but still has some coloring in the banner and the prompt.
![image](https://github.com/user-attachments/assets/3ebb4e28-88ae-4475-b164-bbc14eab78b2)

The banner should be able to be handled in the stdlib perhaps. The prompt is another question. I wonder if the prompts should be ran through ansi-strip or if should be left to the prompt authors? My tendency is to say nushell should remove all coloring. What do you think?

cptpiepmatz commented:
> Sorry to keep hitting you with these "one more things" but I've noticed a couple things.

I have no problem when the iteration cycle is still fast 😄 

> 1. I think we need some text in the [doc_config.nu](https://github.com/nushell/nushell/blob/38694a9850a08f82a9a887c64df55c9f561e4c61/crates/nu-utils/src/default_files/doc_config.nu#L247-L252)

Definitely, will add some

> 2. In my testing, I've noticed that it mostly works but there are some edge cases. NO_COLOR=1 but still has some coloring in the banner and the prompt.

I'm not sure if other commands properly respect `config.use_ansi_coloring`, I made sure the `table` command does, so we should focus testing with that.

> The banner should be able to be handled in the stdlib perhaps. The prompt is another question. I wonder if the prompts should be ran through ansi-strip or if should be left to the prompt authors? My tendency is to say nushell should remove all coloring. What do you think?

I've mentioned this PR in #14194 and I think we should go in the direction to remove ansi colors only from our own output if not explicitely called with `ansi strip`. So if the banner and prompt use some `ansi` commands, they should respect `config.use_ansi_coloring`.

cptpiepmatz commented:
@fdncred I updated `doc_config.nu` to accomodate the new changes.

About the prompt and banner, I think this is out of scope of this PR if they respect the config value or not. The `table` command now definitely respects that value and worked in my testing.

fdncred commented:
I'm fine with landing this as is but I'm not sure we should force 3rd party authors to understand NO_COLOR, FORCE_COLOR, CLI_COLOR. I could see them respecting $env.config.use_ansi_coloring though. But maybe your code is enough with the auto setting. 🤔

fdncred commented:
Let's give it a whirl. Thanks!

cptpiepmatz commented:
> I'm fine with landing this as is but I'm not sure we should force 3rd party authors to understand NO_COLOR, FORCE_COLOR, CLI_COLOR. I could see them respecting $env.config.use_ansi_coloring though. But maybe your code is enough with the auto setting. 🤔

Oh, in the sense to check the config value manually?

132ikl commented:
Maybe the default value of `$env.config.use_ansi_coloring` could be set by the envvars? So:

* `$env.config.use_ansi_coloring` is set
  * Use that
* `$env.config.use_ansi_coloring` is unset (not sure how `CLI_COLOR` is supposed to work)
  * If `FORCE_COLOR` is set: `$env.config.use_ansi_coloring` defaults to true
  * Else, if `NO_COLOR` is set: `$env.config.use_ansi_coloring` defaults to false
  * Else, `$env.config.use_ansi_coloring` defaults to auto

cptpiepmatz commented:
> Maybe the default value of `$env.config.use_ansi_coloring` could be set by the envvars? So:
> 
> * `$env.config.use_ansi_coloring` is set
>   
>   * Use that
> * `$env.config.use_ansi_coloring` is unset (not sure how `CLI_COLOR` is supposed to work)
>   
>   * If `FORCE_COLOR` is set: `$env.config.use_ansi_coloring` defaults to true
>   * Else, if `NO_COLOR` is set: `$env.config.use_ansi_coloring` defaults to false
>   * Else, `$env.config.use_ansi_coloring` defaults to auto

Having the env variables set the config values feels a bit too magic for my taste. Also that would add complexity when adding env variables as you might want to set `NO_COLOR` in the REPL. To make using that feature easiert, we probably need some command the returns the result of the `get` method here.

cptpiepmatz commented:
> I'm fine with landing this as is but I'm not sure we should force 3rd party authors to understand NO_COLOR, FORCE_COLOR, CLI_COLOR. I could see them respecting $env.config.use_ansi_coloring though. But maybe your code is enough with the auto setting. 🤔

I opened #14683 to solve that.
